sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -764,6 +773,184 @@ abstract class TypeCoercionBase {
+        val newChildren = collateToSingleType(checkCastWithoutIndeterminate.children)
+        checkCastWithoutIndeterminate.withNewChildren(newChildren)
+
+      case checkIndeterminate@(_: BinaryExpression | _: In | _: SortOrder)
+        if hasIndeterminate(checkIndeterminate.children
+          .filter(e => hasStringType(e.dataType))
+          .map(e => extractStringType(e.dataType))) =>
+        throw QueryCompilationErrors.indeterminateCollationError()
+
+      case checkImplicitCastInputTypes: ImplicitCastInputTypes
+        if checkImplicitCastInputTypes.children.exists(e => hasStringType(e.dataType))
+          && checkImplicitCastInputTypes.inputTypes.nonEmpty =>
+        val collationId: Int =
+          getOutputCollation(checkImplicitCastInputTypes
+            .children.filter { e => hasStringType(e.dataType) })
+        val children: Seq[Expression] = checkImplicitCastInputTypes
+          .children.zip(checkImplicitCastInputTypes.inputTypes).map {
+            case (e, st) if hasStringType(st) =>
+              castStringType(e, collationId, Some(st)).getOrElse(e)
+            case (e, TypeCollection(types)) if types.exists(hasStringType) =>
+              types.flatMap{ dt =>
+                if (hasStringType(dt)) {
+                  castStringType(e, collationId, Some(dt))
+                } else {
+                  implicitCast(e, dt)
+                }
+              }.headOption.getOrElse(e)
+            case (in, expected) => implicitCast(in, expected).getOrElse(in)
+          }
+        checkImplicitCastInputTypes.withNewChildren(children)
+
+      case checkExpectsInputType: ExpectsInputTypes
+        if checkExpectsInputType.children.exists(e => hasStringType(e.dataType))
+          && checkExpectsInputType.inputTypes.nonEmpty =>
+          val collationId: Int = getOutputCollation(
+            checkExpectsInputType.children.filter {e => hasStringType(e.dataType)})
+          val children: Seq[Expression] = checkExpectsInputType
+            .children.zip(checkExpectsInputType.inputTypes).map {
+            case (st, _) if hasStringType(st.dataType) =>
+              castStringType(st, collationId).getOrElse(st)
+            case (nt, e)
+              if hasStringType(e) && nt.dataType == NullType =>
+              castStringType(nt, collationId, Some(e)).getOrElse(nt)
+            case (nt, e: TypeCollection)
+              if hasStringType(e.defaultConcreteType) && nt.dataType == NullType =>
+              castStringType(nt, collationId, Some(e.defaultConcreteType)).getOrElse(nt)
+            case (in, _) => in
+            }
+          checkExpectsInputType.withNewChildren(children)
+    }
+
+    def shouldCast(types: Seq[Expression]): Boolean = {
+      types.filter(e => hasStringType(e.dataType))
+        .map(e => extractStringType(e.dataType).collationId).distinct.size > 1
+    }
+
+    /**
+     * Whether the data type contains StringType.
+     */
+    @tailrec
```

Comment:
You have `existsRecursively` which we use for similar purpose in QP side. E.g.
```scala
  def isBinaryStable(dataType: DataType): Boolean = dataType.existsRecursively {
    case st: StringType => CollationFactory.fetchCollation(st.collationId).isBinaryCollation
    case _ => true
  }
```

Reply:
This is not possible for now. I think this PR is already too big and if I change it to this I need to deal with StructType and MapType. I suggest we leave those two for follow-up PR.